pkg/controller/history/controller_history.go

```diff
@@ -251,6 +251,7 @@ func (rh *realHistory) CreateControllerRevision(parent metav1.Object, revision *
 		hash := HashControllerRevision(revision, collisionCount)
 		// Update the revisions name and labels
 		clone.Name = ControllerRevisionName(parent.GetName(), hash)
+		clone.Labels[ControllerRevisionHashLabel] = strconv.FormatInt(int64(hash), 10)
 		created, err := rh.client.AppsV1beta1().ControllerRevisions(parent.GetNamespace()).Create(clone)
 		if errors.IsAlreadyExists(err) {
 			*collisionCount++
```

Comment:
collisionCount shouldn't be blindly increased when there's name collision. Because controller uses a cache store, the name collision could be due to a controllerrevision created but not yet observed by the controller. 

The right way to handle it is to first check if the controllerrevision with the same name has the same content as the one we're going to create, like how DaemonSet controller handles it:

https://github.com/kubernetes/kubernetes/blob/296bc649241a20294da11000c6c6e82cca280c81/pkg/controller/daemon/update.go#L333-L366

Reply:
Thanks for your information, update